### PR TITLE
First draft to support lvm storage (topolvm)

### PR DESCRIPTION
**Pull request by dhenkel92**
In order to support more fine grained local storage, we are currently investigating different solutions like [topolvm](https://github.com/topolvm/topolvm) or [openebs](https://github.com/openebs/openebs).

One of the most important requirement is that the new solution should also support autoscaling. So, this is the first draft to implement a pretty straight forward way to tell the autoscaler all the information it needs to figure out which node group can support what storage requirements.

### CODE REVIEW

File: cluster-autoscaler/processors/datadog/pods/transform_local_data_test.go [hunk 1]
```diff
@@ -92,6 +98,26 @@ func TestTransformLocalDataProcess(t *testing.T) {
...
+			},
+			[]*corev1.Pod{buildPod("pod1", testTopolvmResources, testTopolvmResources, "pvc-1")},
+		},
+
+		{
+			"one pvc will override the other",
+			[]*corev1.Pod{buildPod("pod1", testEmptyResources, testEmptyResources, "pvc-1", "pvc-2")},
+			[]*corev1.PersistentVolumeClaim{
+				buildPVCWithStorage("pvc-1", storageClassNameTopolvm, "100Gi"),
+				buildPVC("pvc-2", storageClassNameLocal),
```
Comment (dhenkel92) [inline cluster-autoscaler/processors/datadog/pods/transform_local_data_test.go:117]: Should it be possible to have multiple PVC's?

**Comment by dhenkel92** [comment]
Do we want to keep the local storage label to indicate that there is storage?
```
// DatadogLocalStorageLabel is "true" on nodes offering local storage
DatadogLocalStorageLabel = "nodegroups.datadoghq.com/local-storage"
```

File: cluster-autoscaler/processors/datadog/common/common.go [hunk 2]
```diff
@@ -61,7 +68,22 @@ func SetNodeLocalDataResource(nodeInfo *schedulerframework.NodeInfo) {
 	if node.Status.Capacity == nil {
 		node.Status.Capacity = apiv1.ResourceList{}
 	}
-	node.Status.Capacity[DatadogLocalDataResource] = DatadogLocalDataQuantity.DeepCopy()
-	node.Status.Allocatable[DatadogLocalDataResource] = DatadogLocalDataQuantity.DeepCopy()
+
+	provisioner, _ := node.Labels[DatadogLocalStorageProvisionerLabel]
+	switch provisioner {
+	case "topolvm":
```
Comment (antoine-gaillard) [inline cluster-autoscaler/processors/datadog/common/common.go:74]: I think we can either add "openebs-lvm" as well here or pass down the provisioning mode field to the node themselves, but we can do that in another PR

File: cluster-autoscaler/processors/datadog/pods/transform_local_data_test.go (hunk 1 above)
Comment (antoine-gaillard) [reply cluster-autoscaler/processors/datadog/pods/transform_local_data_test.go:117]: we should be able to have 2 lvm pvc on the same node, we can't have "legacy" local-data and lvm at the same time on the nodes though

Comment (dhenkel92) [reply cluster-autoscaler/processors/datadog/pods/transform_local_data_test.go:117]: Sorry I've added the comment at the worst place. 
The question was more like if one Pod can have multiple topolvm volume claims simultaneously because this would mean we have to add up the total capacity needed. Otherwise, the latter will override the first.

File: cluster-autoscaler/processors/datadog/common/common.go (hunk 2 above)
Comment (dhenkel92) [reply cluster-autoscaler/processors/datadog/common/common.go:74]: I'll do 👍

Comment (dhenkel92) [reply cluster-autoscaler/processors/datadog/common/common.go:74]: https://github.com/DataDog/autoscaler/pull/53

File: cluster-autoscaler/processors/datadog/pods/transform_local_data_test.go (hunk 1 above)
Comment (antoine-gaillard) [reply cluster-autoscaler/processors/datadog/pods/transform_local_data_test.go:117]: hmm good question, I don't think we want to support that right now but there might be use cases

**Comment by antoine-gaillard** [comment]
> Do we want to keep the local storage label to indicate that there is storage?
> 
> ```
> // DatadogLocalStorageLabel is "true" on nodes offering local storage
> DatadogLocalStorageLabel = "nodegroups.datadoghq.com/local-storage"
> ```



> Do we want to keep the local storage label to indicate that there is storage?
> 
> ```
> // DatadogLocalStorageLabel is "true" on nodes offering local storage
> DatadogLocalStorageLabel = "nodegroups.datadoghq.com/local-storage"
> ```

Probably not since we use a custom storage class and the capacity label to schedule stuff but maybe I'm missing something, @bpineau might know better

File: cluster-autoscaler/processors/datadog/common/common.go [hunk 3]
```diff
@@ -61,7 +68,22 @@ func SetNodeLocalDataResource(nodeInfo *schedulerframework.NodeInfo) {
...
+		capacity, _ := node.Labels[DatadogInitialStorageCapacityLabel]
+		capacityResource, err := resource.ParseQuantity(capacity)
+		if err == nil {
+			node.Status.Capacity[DatadogLocalDataResource] = capacityResource.DeepCopy()
+			node.Status.Allocatable[DatadogLocalDataResource] = capacityResource.DeepCopy()
+		} else {
+			klog.Warningf("failed to attach storage information to node (%s): %v", node.Name, err)
```
Comment (bpineau) [inline cluster-autoscaler/processors/datadog/common/common.go:81]: Can we add the bogus/unparsable capacity string to that log line, to ease diagnostics?

File: cluster-autoscaler/processors/datadog/common/common.go [hunk 4]
```diff
@@ -61,7 +68,22 @@ func SetNodeLocalDataResource(nodeInfo *schedulerframework.NodeInfo) {
...
+		}
+	default:
```
Comment (bpineau) [inline cluster-autoscaler/processors/datadog/common/common.go:83]: Can we spell the expected cases for the later reader? (so here `storageClassNameLoca:l` case, and maybe a default: log("what's going on here?") perhaps)

File: cluster-autoscaler/processors/datadog/pods/transform_local_data.go [hunk 5]
```diff
@@ -114,15 +119,37 @@ func (p *transformLocalData) Process(ctx *context.AutoscalingContext, pods []*ap
...
+			switch *pvc.Spec.StorageClassName {
+			case storageClassNameTopolvm:
+				if storage, ok := pvc.Spec.Resources.Requests["storage"]; ok {
+					po.Spec.Containers[0].Resources.Requests[common.DatadogLocalDataResource] = storage.DeepCopy()
+					po.Spec.Containers[0].Resources.Limits[common.DatadogLocalDataResource] = storage.DeepCopy()
+				} else {
+					klog.Warningf("ignoring pvc as it does not have storage request information")
+					volumes = append(volumes, vol)
+				}
+			default:
```
Comment (bpineau) [inline cluster-autoscaler/processors/datadog/pods/transform_local_data.go:131]: also here?

**Review by bpineau — COMMENTED** [review]
lgtm, just a couple minor comments.
Can you prefix your commit with "[local] " ? (to ease rebase / distinguish backports vs. local-only changes)

**Comment by dhenkel92** [comment]
I've merged the openebs branch and squashed everything into one commit that is prefixed with `[local]`

**Review by bpineau — APPROVED** [review]

**Review by antoine-gaillard — APPROVED** [review]

**Comment by adrienjt** [comment]
> > Do we want to keep the local storage label to indicate that there is storage?
> > ```
> > // DatadogLocalStorageLabel is "true" on nodes offering local storage
> > DatadogLocalStorageLabel = "nodegroups.datadoghq.com/local-storage"
> > ```
> 
> Probably not since we use a custom storage class and the capacity label to schedule stuff but maybe I'm missing something, @bpineau might know better

The local-storage label is used in other places, where the meaning is "whatever the provisioner", so we shouldn't remove it, for example:
- https://github.com/DataDog/k8s-node-metadata-controller/blob/e2007eb035a6b6809d47918bb90e54dde8210e15/pkg/controllers/node_controller.go#L175 (Antoine's work BTW)
- https://github.com/DataDog/k8s-platform-resources/blob/f30ae8af5f05b0cacd1db3d326a3c812373bafa6/k8s/k8s-platform-draino/values.yaml#L101
- ...

So we shouldn't remove it.

File: cluster-autoscaler/processors/datadog/common/common.go [hunk 6]
```diff
@@ -61,7 +74,30 @@ func SetNodeLocalDataResource(nodeInfo *schedulerframework.NodeInfo) {
...
 		node.Status.Capacity = apiv1.ResourceList{}
 	}
-	node.Status.Capacity[DatadogLocalDataResource] = DatadogLocalDataQuantity.DeepCopy()
-	node.Status.Allocatable[DatadogLocalDataResource] = DatadogLocalDataQuantity.DeepCopy()
+
+	provisioner, _ := node.Labels[DatadogLocalStorageProvisionerLabel]
+	switch provisioner {
+	case "openebs-lvm":
+		fallthrough
+	case "topolvm":
```
[the remark this answers is not in the document]
Comment (dhenkel92) [reply cluster-autoscaler/processors/datadog/common/common.go:82]: 👍 I wasn't aware of this syntactic sugar.

And yes agree, they should be constants.
